server/backends/chunkstore/chunkstore.go

```diff
@@ -388,164 +394,185 @@ type ChunkstoreWriter struct {
+		select {
+		case req, open = <-l.writeChannel:
+			if req != nil {
+				// We received a write request; update the context, chunk, and tail.
+				ctx = req.ctx
+				if req.Chunk == nil {
+					// nil Write is a manual flush
+					flushTime = time.Unix(0, 0)
+				}
 				if req.VolatileTail != nil {
-					l.volatileTail = req.VolatileTail
+					// nil Tail means don't overwrite the tail. This accompanies manual
+					// flushes and no-op writes (to get updated chunkIndex from potential
+					// automatic flushes).
+					volatileTail = req.VolatileTail
+				}
+				chunk = append(chunk, req.Chunk...)
+				lastWriteSize = len(req.Chunk)
+				if req.Close {
+					open = false
 				}
-				l.chunk = append(l.chunk, req.Chunk...)
-				l.lastWriteSize = len(req.Chunk)
 			}
+		case <-time.After(time.Until(flushTime)):
+			// We timed out waiting for a write request. Flush the contents of the
+			// chunk to disk.
+			timeout = true
+		case <-ctx.Done():
+			open = false
+
+			// The context was canceled before the file was closed; extend the life of
+			// the context so we can flush the contents of the chunk and tail to disk.
+			var cancel func()
+			ctx, cancel = background.ExtendContextForFinalization(ctx, time.Second*10)
+			defer cancel()
 		}
-		return req
-	case <-time.After(time.Until(l.flushTime)):
-		l.timeout = true
-		return nil
-	case <-l.ctx.Done():
-		l.open = false
-		return nil
-	}
-}
 
-func (l *writeLoop) run() {
-	l.flushTime = time.Now().Add(year)
-	l.open = true
-	for l.open {
-		// Get the read request and update loop for this iteration.
-		req := l.readFromWriteChannel()
-		if !l.open {
+		if !open {
 			// We are closing the file, so any tail must be written lest it be lost.
-			l.chunk = append(l.chunk, l.volatileTail...)
-			l.lastWriteSize = len(l.volatileTail)
+			chunk = append(chunk, volatileTail...)
+			lastWriteSize = len(volatileTail)
+			volatileTail = []byte{}
 		}
 		var err error
-		for err == nil && len(l.chunk) >= l.writeBlockSize {
+		for err == nil && len(chunk) >= l.writeBlockSize {
 			// Write blocks until we have fewer than writeBlockSize bytes remaining
 			// or we encounter a write error.
-			err = l.write()
+			var n int
+			n, err = l.write(ctx, chunk, chunkIndex, lastWriteSize)
+			if err == nil {
+				bytesFlushed += n
+				chunk = chunk[n:]
+				chunkIndex++
+			}
 		}
 		// We do not attempt to write if we have already encountered a write error
 		// in this iteration of the loop, and instead we simply move on.
 		// If the file is empty and we are closing it, we must write an empty chunk.
 		// Otherwise, if we have data cached to write and we are either flushing or
 		// closing the file, we need to write a chunk.
-		if err == nil && ((!l.open && l.chunkIndex == 0) || (len(l.chunk) > 0 && (time.Now().After(l.flushTime) || !l.open))) {
-			l.write()
+		if err == nil && ((!open && chunkIndex == 0) || (len(chunk) > 0 && (time.Now().After(flushTime) || !open))) {
```

Comment:
I don't understand the `chunkIndex == 0` part of this conditional, mind adding that to the comment?

Also, should `len(chunk) > 0` be required for all parts of the conditional? i.e. `err == nil && len(chunk) > 0 && ( ... )`

Reply:
I tried rewriting it, maybe it's easier to understand now?